src/main/java/com/blamejared/crafttweaker/impl/predicate/LocationPredicate.java

```diff
@@ -0,0 +1,200 @@
+    public LocationPredicate withDimension(final ResourceLocation dimension) {
+        this.dimension = dimension;
+        return this;
+    }
+
+    @ZenCodeType.Method
+    public LocationPredicate withFeatureName(final String feature) {
+        this.feature = feature;
+        return this;
+    }
+
+    @SuppressWarnings("SpellCheckingInspection")
+    @ZenCodeType.Method
+    public LocationPredicate withBiomeName(final ResourceLocation biome) {
+        this.biome = biome;
+        return this;
+    }
+
+    @ZenCodeType.Method
+    public LocationPredicate withCampfireBelow() {
+        this.aboveCampfire = TriState.TRUE;
+        return this;
+    }
+
+    @ZenCodeType.Method
+    public LocationPredicate withoutCampfireBelow() {
+        this.aboveCampfire = TriState.FALSE;
+        return this;
+    }
+
+    @ZenCodeType.Method
+    public LocationPredicate withLightPredicate(final Consumer<LightPredicate> builder) {
+        final LightPredicate predicate = new LightPredicate();
+        builder.accept(predicate);
+        this.lightLevel = predicate;
+        return this;
+    }
+
+    @ZenCodeType.Method
+    public LocationPredicate withBlockPredicate(final Consumer<BlockPredicate> builder) {
+        final BlockPredicate predicate = new BlockPredicate();
+        builder.accept(predicate);
+        this.block = predicate;
+        return this;
+    }
+
+    @ZenCodeType.Method
+    public LocationPredicate withFluidPredicate(final Consumer<FluidPredicate> builder) {
+        final FluidPredicate predicate = new FluidPredicate();
+        builder.accept(predicate);
+        this.fluid = predicate;
+        return this;
+    }
+
+    @Override
+    public boolean isAny() {
+        return this.x.isAny() && this.y.isAny() && this.z.isAny() && this.dimension == null && this.feature == null
+                && this.biome == null && this.aboveCampfire.isUnset() && this.lightLevel.isAny() && this.block.isAny()
+                && this.fluid.isAny();
+    }
+
+    @Override
+    public net.minecraft.advancements.criterion.LocationPredicate toVanilla() {
+        return new net.minecraft.advancements.criterion.LocationPredicate(
+                this.x.toVanillaPredicate(),
+                this.y.toVanillaPredicate(),
+                this.z.toVanillaPredicate(),
+                this.toVanilla(ForgeRegistries.BIOMES, this.biome),
+                this.feature == null? null : Structure.NAME_STRUCTURE_BIMAP.get(this.feature),
+                this.toVanilla(Registry.WORLD_KEY, this.dimension),
+                this.aboveCampfire.toBoolean(),
+                this.lightLevel.toVanillaPredicate(),
+                this.block.toVanillaPredicate(),
+                this.fluid.toVanillaPredicate()
+        );
+    }
+
+    private <T> RegistryKey<T> toVanilla(final RegistryKey<? extends Registry<T>> registry, final ResourceLocation name) {
+        if (name == null) return null;
+        return RegistryKey.getOrCreateKey(registry, name);
```

Comment:
Is getOrCreateKey the proper method here?
Especially the `OrCreate` part?

Reply:
Yes: `RegistryKey`s are created and cached, so there is no way to obtain one without consulting a cache nor without creating it if it doesn't exist